src/Workspaces/SharedUtilitiesAndExtensions/Workspace/CSharp/LanguageServices/CSharpTypeInferenceService.TypeInferrer.cs

```diff
@@ -1234,6 +1236,86 @@ private IEnumerable<TypeInferenceInfo> InferTypeInPropertyDeclaration(PropertyDe
+                        SemanticModel.Compilation, collectionType, out var elementTypes) &&
+                    elementTypes is [var elementType, ..])
+                {
+                    return [new(elementType)];
+                }
+
+                // If that fails, see if we can figure out from one of our siblings.
+                foreach (var element in collectionExpression.Elements)
+                {
+                    if (element != expressionElement && element is ExpressionElementSyntax siblingElement)
+                    {
+                        var types = GetTypes(siblingElement.Expression, objectAsDefault: false);
+                        if (types.Any())
+                            return types;
+                    }
+                }
+            }
+
+            return [];
+        }
+
+        private IEnumerable<TypeInferenceInfo> InferTypeInKeyValuePairElement(
+            KeyValuePairElementSyntax keyValuePairElement, ExpressionSyntax expression)
+        {
+            var isKey = expression == keyValuePairElement.KeyExpression;
+            var types = InferTypes();
+            return types.Select(t => new TypeInferenceInfo(t));
+
+            IEnumerable<ITypeSymbol> InferTypes()
+            {
+                if (keyValuePairElement.Parent is CollectionExpressionSyntax collectionExpression)
+                {
+                    var collectionType = SemanticModel.GetTypeInfo(collectionExpression, CancellationToken).ConvertedType;
+
+                    // Try to figure out the type based on the type of hte collection itself.
+                    if (CollectionExpressionExtensions.IsConstructibleCollectionType(
+                            SemanticModel.Compilation, collectionType, out var elementTypes) &&
+                        elementTypes is [var elementType, ..])
+                    {
+                        return PickKeyOrValueType(elementType);
+                    }
+
+                    // If that fails, see if we can figure out from one of our siblings.
+                    foreach (var element in collectionExpression.Elements)
+                    {
+                        if (element != keyValuePairElement && element is KeyValuePairElementSyntax siblingElement)
+                        {
+                            var types = GetTypes(isKey ? siblingElement.KeyExpression : siblingElement.ValueExpression, objectAsDefault: false);
+                            if (types.Any())
+                                return types.SelectMany(i => PickKeyOrValueType(i.InferredType));
+                        }
+                    }
+                }
+
+                return [];
+            }
+
+            IEnumerable<ITypeSymbol> PickKeyOrValueType(ITypeSymbol typeSymbol)
+            {
+                if (typeSymbol is INamedTypeSymbol
+                    {
+                        Name: nameof(KeyValuePair<int, int>),
```

Comment:
```suggestion
                        Name: nameof(KeyValuePair<,>),
```

Or can we not use that for some reason?

Reply:
we can't use that just yet.  need to have a public release.